ui/components-react/src/test/tree/controlled/TreeEventDispatcher.test.ts

```diff
@@ -3,604 +3,823 @@
-          selectedNodeIds: selection,
-        });
-        expect(spy).to.be.called;
-
-        const spyArgs = spy.args[0][0] as TreeSelectionReplacementEventArgs;
-        const results = await extractSequence(rxjsFrom(spyArgs.replacements));
-        expect(results).to.not.be.empty;
-        const selectionChange = results[0];
-        expect(selectionChange.selectedNodeItems).to.be.deep.eq([
-          deselectedNodes[0].item,
+      dispatcher.onNodeClicked("A", { shiftKey: true } as any);
+      expect(treeEvents.onSelectionReplaced).to.be.calledTwice;
+      const [args1] = treeEvents.onSelectionReplaced.args[1];
+      expect(await extractSequence(toRxjsObservable(args1.replacements)))
+        .to.have.lengthOf(1)
+        .and.containSubset([
+          {
+            selectedNodeItems: [{ id: "A" }],
+          },
         ]);
-      });
     });
   });
 
-  describe("onNodeCheckboxClicked", () => {
-    it("changes state for clicked node", async () => {
-      const expectedAffectedNodeItems = [deselectedNodes[0].item];
+  describe("drag selection", () => {
+    it("selects nodes", async () => {
+      const { dispatcher } = setupTreeEventDispatcher(
+        SelectionMode.Multiple,
+        (model) => {
+          model.setChildren(
+            undefined,
+            [
+              createTreeNodeInput("A"),
+              createTreeNodeInput("B"),
+              createTreeNodeInput("C"),
+            ],
+            0
+          );
+        }
+      );
+
+      dispatcher.onNodeMouseDown("A");
+
+      expect(treeEvents.onSelectionModified).to.be.calledOnce;
+      const [args0] = treeEvents.onSelectionModified.args[0];
+      const result0 = startExtractingSequence(
+        toRxjsObservable(args0.modifications)
+      );
 
-      const spy = sinon.spy();
-      treeEventsMock.setup((x) => x.onCheckboxStateChanged).returns(() => spy);
+      dispatcher.onNodeMouseMove("B");
+      dispatcher.onNodeMouseMove("C");
+      fireEvent.mouseUp(window);
 
-      dispatcher.onNodeCheckboxClicked(deselectedNodes[0].id, CheckBoxState.On);
+      await result0.waitForComplete;
+      expect(result0.current.sequence)
+        .to.have.lengthOf(2)
+        .and.containSubset([
+          {
+            selectedNodeItems: [{ id: "A" }, { id: "B" }],
+            deselectedNodeItems: [],
+          },
+          {
+            selectedNodeItems: [{ id: "C" }],
+            deselectedNodeItems: [],
+          },
+        ]);
 
-      expect(spy).to.be.calledOnce;
-      const changes = spy.args[0][0] as TreeCheckboxStateChangeEventArgs;
-      const results = await extractSequence(rxjsFrom(changes.stateChanges));
-      expect(results).to.not.be.empty;
-      const affectedNodeItems = results[0].map((change) => change.nodeItem);
-      expect(affectedNodeItems).to.be.deep.eq(expectedAffectedNodeItems);
+      expect(treeEvents.onSelectionModified).to.be.calledTwice;
+      const [args1] = treeEvents.onSelectionModified.args[1];
```

Comment:
what causes `onSelectionModified` to be called the second time?

Reply:
I see similar situations in other tests as well - could you please just describe when/why `onSelectionModified` are fired on drag selection?

Reply:
When drag selection is in progress `SelectionHandler` is accumulating all changes in `BatchSelectionOperation` https://github.com/iTwin/appui/blob/71ab57c9416bb598688d13f47bff3563489d30ac/ui/components-react/src/components-react/common/selection/SelectionHandler.ts#L461
and when drag selection is finished items are `selected/deselected` based on changes accumulated in that operation https://github.com/iTwin/appui/blob/71ab57c9416bb598688d13f47bff3563489d30ac/ui/components-react/src/components-react/common/selection/SelectionHandler.ts#L395

I guess this stands as source of truth of what was `selected/deselect` during drag operation. I see it makes separate calls for selected and deselected items so in other test `onSelectedModified` ends up called three times.

Reply:
thanks, I think I understand this now (comments that you added to the tests are really helpful)